### PR TITLE
Compute ParametricGeometry normals from tangent vectors

### PR DESCRIPTION
**Pull request by jaxry**
The ParametricGeometry class generates vertex normals as usual from the geometry's faces. However, this method leads to incorrectly computed normals for closed parametric surfaces. Where two ends of a surface meet, a seam is created since the positions along the edge are shared, but the indices of the geometry aren't.

For example, here you can see two seams on this torus created using ParametricGeometry:
normals from faces

This PR attempts to solve the problem. These seams are removed if you calculate the normals from the parametric equation itself.
normals from derivatives
By approximating tangent vectors for each vertex via the finite difference method and taking the cross product of the vectors, a correct surface normal is retrieved.

I have a few concerns about my code.
* The `computeVertexNormals` method remains untouched, so if this method is called, the normals will be computed from the geometry faces again. The method could be overridden in ParametricGeometry and made to compute normals from the parametric equation, but the overridden method would contain a lot of duplicated code, since the method would still function very similarly. I'm not sure how to proceed.

### CODE REVIEW

File: src/geometries/ParametricGeometry.js [hunk 1]
```diff
@@ -72,6 +77,16 @@ function ParametricBufferGeometry( func, slices, stacks ) {
 			var p = func( u, v );
 			vertices.push( p.x, p.y, p.z );
 
+			// approximate tangent plane vectors via central difference
+
+			pu.subVectors( func( u + EPS, v ), func( u - EPS, v ) );
```
Comment (jaxry) [inline src/geometries/ParametricGeometry.js]: Should both `Vector3`'s returned by `func` be `.clone`ed here in case the objects aren't unique? Or is there a better way to do this?

File: src/geometries/ParametricGeometry.js [hunk 2]
```diff
@@ -72,6 +77,16 @@ function ParametricBufferGeometry( func, slices, stacks ) {
...
+			pv.subVectors( func( u, v + EPS ), func( u, v - EPS ) );
```
Comment (jaxry) [inline src/geometries/ParametricGeometry.js]: Should both `Vector3`'s returned by `func` be `.clone`ed here in case both objects are actually the same? Or is there a better way to do this?

**Comment by Mugen87** [comment]
> The computeVertexNormals method remains untouched, so if this method is called, the normals will be computed from the geometry faces again. The method could be overridden in ParametricGeometry and made to compute normals from the parametric equation, but the overridden method would contain a lot of duplicated code, since the method would still function very similarly. I'm not sure how to proceed.

There is no need to override `computeVertexNormals`. It's not intended that users call `computeVertexNormals` after the geometry was generated.

File: src/geometries/ParametricGeometry.js (hunk 2 above)
Comment (Mugen87) [inline src/geometries/ParametricGeometry.js]: `func` normally returns a new instance of `Vector3`, see https://github.com/mrdoob/three.js/blob/dev/examples/js/ParametricGeometries.js

**Comment by Mugen87** [comment]
Looks good! :+1:

File: src/geometries/ParametricGeometry.js (hunk 1 above)
Comment (WestLangley) [inline src/geometries/ParametricGeometry.js]: Is `func( u, v )` guaranteed to be well-defined for values of `u, v` outside [ 0, 1 ] ?

File: src/geometries/ParametricGeometry.js [hunk 3]
```diff
@@ -72,6 +77,16 @@ function ParametricBufferGeometry( func, slices, stacks ) {
...
+
+			// cross product of tangent plane vectors returns surface normal
+
+			normal.crossVectors( pu, pv );
```
Comment (WestLangley) [inline src/geometries/ParametricGeometry.js]:
```javascript
normal.crossVectors( pu, pv ).normalize();
```

File: src/geometries/ParametricGeometry.js [hunk 4]
```diff
@@ -102,11 +117,10 @@ function ParametricBufferGeometry( func, slices, stacks ) {
 
 	this.setIndex( indices );
 	this.addAttribute( 'position', new Float32BufferAttribute( vertices, 3 ) );
+	this.addAttribute( 'normal', new Float32BufferAttribute( normals, 3 ) );
 	this.addAttribute( 'uv', new Float32BufferAttribute( uvs, 2 ) );
 
-	// generate normals
-
-	this.computeVertexNormals();
+	this.normalizeNormals();
```
Comment (WestLangley) [inline src/geometries/ParametricGeometry.js]: not needed

File: src/geometries/ParametricGeometry.js (hunk 2 above)
Comment (WestLangley) [inline src/geometries/ParametricGeometry.js]: @Mugen87 Actually, there should be no reason to instantiate `4 * slices * stacks` `Vector3`'s when using central differences. We could add an `optionalTarget` to `func()` and reuse just two instances.
```javascript
var result = optionalTarget || new THREE.Vector3();
...
return result.set( x, y, z );
```

**Comment by WestLangley** [comment]
Yes, `computeVertexNormals()` is a heuristic and should not be called by constructors when the normals can be computed analytically or by finite differences.

This is a good change as long as `func()` is not called with values that are out-of-range.

File: src/geometries/ParametricGeometry.js (hunk 3 above)
Comment (Mugen87) [inline src/geometries/ParametricGeometry.js]: 👍

File: src/geometries/ParametricGeometry.js (hunk 1 above)
Comment (Mugen87) [inline src/geometries/ParametricGeometry.js]: That's a good point! According to the implementation of `func`, values outside [ 0, 1 ] can be problematic.

File: src/geometries/ParametricGeometry.js (hunk 2 above)
Comment (Mugen87) [inline src/geometries/ParametricGeometry.js]: Sounds reasonable. We can also adjust some entities in `THREE.ParametricGeometries` so they use the `optionalTarget`. But let's do this in a separate PR, okay?

@jaxry Are you interested in doing this 😉 ?

File: src/geometries/ParametricGeometry.js (hunk 1 above)
Comment (WestLangley) [inline src/geometries/ParametricGeometry.js]: Then, use non-centered differences with a special case for the edge. By reusing `p`, only two additional calls are needed.

File: src/geometries/ParametricGeometry.js (hunk 2 above)
Comment (WestLangley) [inline src/geometries/ParametricGeometry.js]: Yes. Separate PR.

**Comment by jaxry** [comment]
Thanks for the feedback. I implemented your suggestions.

**Comment by mrdoob** [comment]
Thanks!
